### PR TITLE
cryptenroll: Implement support for unlocking via FIDO2 tokens

### PR DESCRIPTION
**Pull request by PeterCxy**
This allows FIDO2 users to wipe out password slots and still be able to
enroll new key slots via systemd-cryptenroll. Note that when the user
wants to both unlock with a FIDO2 token and enroll a new FIDO2 token,
they cannot be set to automatic discovery. This is to safeguard against
confusion, because there will be multiple tokens connected to the system
when doing so -- and we require users to explicitly confirm which one to
use for unlocking and which one to use for enrollment.

Addresses #20230 for the FIDO2 case.

### CODE REVIEW

File: src/cryptenroll/cryptenroll.h [hunk 1]
```diff
@@ -13,6 +13,12 @@ typedef enum EnrollType {
         _ENROLL_TYPE_INVALID = -EINVAL,
 } EnrollType;
 
+typedef enum UnlockType {
+        UNLOCK_PASSWORD,
+        UNLOCK_KEYFILE,
+        UNLOCK_FIDO2,
```
Comment (bluca) [inline src/cryptenroll/cryptenroll.h:19]: _MAX and _INVALID

File: src/cryptenroll/cryptenroll-fido2.c [hunk 2]
```diff
@@ -1,12 +1,55 @@
...
 #include "json.h"
 #include "libfido2-util.h"
 #include "memory-util.h"
 #include "random-util.h"
 
+int prepare_luks_fido2(
+                struct crypt_device *cd,
+                const char *device,
+                void *ret_vk,
+                size_t *ret_vks) {
```
Comment (bluca) [inline src/cryptenroll/cryptenroll-fido2.c:17]: empty line after `{`

File: src/cryptenroll/cryptenroll.c [hunk 3]
```diff
@@ -474,6 +508,39 @@ static int check_for_homed(struct crypt_device *cd) {
         return 0;
 }
 
+static int prepare_luks_keyfile(
+                struct crypt_device *cd,
+                void *ret_vk,
+                size_t *ret_vks) {
```
Comment (bluca) [inline src/cryptenroll/cryptenroll.c:521]: empty line after `{`

File: src/shared/cryptsetup-fido2.c [hunk 4]
```diff
@@ -247,12 +243,12 @@ int acquire_fido2_key_auto(
                         required |= FIDO2ENROLL_UV_OMIT; /* compat with 248 */
 
                 ret = acquire_fido2_key(
-                                name,
-                                friendly_name,
+                                NULL,
```
Comment (bluca) [inline src/shared/cryptsetup-fido2.c]: mark the NULL/0/true/false parameters with their names - ie, `/* volume_name= */ NULL,` here and elsewhere

**Review by bluca — CHANGES_REQUESTED** [review]
some style nits

File: src/cryptenroll/cryptenroll.h (hunk 1 above)
Comment (PeterCxy) [reply src/cryptenroll/cryptenroll.h:19]: Do we need _INVALID here though? Because `arg_unlock_type` defaults to `UNLOCK_PASSWORD` anyway, and _INVALID will never be used.

File: src/shared/cryptsetup-fido2.c (hunk 4 above)
Comment (PeterCxy) [reply src/shared/cryptsetup-fido2.c]: Added comments for argument names, but I didn't change the ones in `prepare_luks_keyfile()` because code there is copied as-is from the original `prepare_luks()` function.

File: src/cryptenroll/cryptenroll-fido2.c [hunk 5]
```diff
@@ -1,12 +1,56 @@
...
+        if (r < 0)
+                return log_oom();
+
+        r = crypt_volume_key_get(
+                        cd,
+                        CRYPT_ANY_SLOT,
+                        ret_vk,
+                        ret_vks,
+                        passphrase,
+                        strlen(passphrase));
```
Comment (keszybz) [inline src/cryptenroll/cryptenroll-fido2.c]: base64mem returns the length… It doesn't matter *greatly*, but since we already have the data, it's seems nicer to avoid the slow and pointless strlen call.

File: src/cryptenroll/cryptenroll.c [hunk 6]
```diff
@@ -253,7 +260,29 @@ static int parse_argv(int argc, char *argv[]) {
...
+                        arg_unlock_type = UNLOCK_KEYFILE;
+                        break;
+
+                case ARG_UNLOCK_FIDO2_DEVICE: {
+                        _cleanup_free_ char *device = NULL;
+
+                        if (streq(optarg, "auto")) {
+                                if (arg_enroll_type == ENROLL_FIDO2)
+                                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                               "When enrolling a new FIDO2 device, `--unlock-fido2-device=auto` is unsupported. "
```
Comment (keszybz) [inline src/cryptenroll/cryptenroll.c]: Please don't use backticks. They **only** make sense in markdown, and are not good for quoting in other places. They are also special to the shell. Just use normal single quotes.

File: src/cryptenroll/cryptenroll.c [hunk 7]
```diff
@@ -253,7 +260,29 @@ static int parse_argv(int argc, char *argv[]) {
...
+                                                               "Please specify a device to use for unlocking.");
+                        } else {
+                                device = strdup(optarg);
+                                if (!device)
+                                        return log_oom();
+
+                                arg_unlock_fido2_device = TAKE_PTR(device);
```
Comment (keszybz) [inline src/cryptenroll/cryptenroll.c]: This creates a potential memleak: if `--unlock-fido2-device` is specified more than once, the first `strdup` will be lost. But I think you don't need the strdup at all… Just use a `static const char *` and assing `optarg` directly to it.

File: src/cryptenroll/cryptenroll.c [hunk 8]
```diff
@@ -253,7 +260,29 @@ static int parse_argv(int argc, char *argv[]) {
...
                                 return r;
+
+                        arg_unlock_type = UNLOCK_KEYFILE;
+                        break;
+
+                case ARG_UNLOCK_FIDO2_DEVICE: {
+                        _cleanup_free_ char *device = NULL;
+
+                        if (streq(optarg, "auto")) {
+                                if (arg_enroll_type == ENROLL_FIDO2)
```
Comment (keszybz) [inline src/cryptenroll/cryptenroll.c]: IIUC, `arg_enroll_type` is set by another option. For this check to work, the user needs to use *that* option first and *this* option second. If they do it in opposite order, the check will be skipped. I think this check should be moved to end of this function, after all the options have been processed and we know the final values.

**Review by keszybz — CHANGES_REQUESTED** [review]
Looks very nice. I'm surprised how little new code is required. I didn't test it at all though.

File: src/cryptenroll/cryptenroll.c (hunk 8 above)
Comment (PeterCxy) [reply src/cryptenroll/cryptenroll.c]: This is why the check is done in both the case for `ARG_UNLOCK_FIDO2_DEVICE` and for `ARG_FIDO2_DEVICE`. But yes I guess it would be better to do the check at once after the loop.

File: src/cryptenroll/cryptenroll.c (hunk 7 above)
Comment (PeterCxy) [reply src/cryptenroll/cryptenroll.c]: I kept the `strdup` for consistency with other cases, but I added a guard before all the unlock options just like the enrolling ones to prevent these parameters being passed multiple times.

**Review by bluca — APPROVED** [review]

File: src/shared/cryptsetup-fido2.c [hunk 9]
```diff
@@ -129,12 +131,7 @@ int acquire_fido2_key(
 
 int acquire_fido2_key_auto(
                 struct crypt_device *cd,
-                const char *name,
-                const char *friendly_name,
```
Comment (poettering) [inline src/shared/cryptsetup-fido2.c:133]: Please do not drop `friendly_name` here. We really should include that in the PIN prompt. The fact that we currently do not is an omission we should fix.

i.e. instead of dropping this parameter please change the ask_password_auto() line above to include it in the question, so that it is clear for which volume we need the token.

Comment (poettering) [reply src/shared/cryptsetup-fido2.c:133]: and the `name` should not be dropped either, but passed in the `id` parameter of ask_password_auto() (i.e. the third one, which is currently NULL), so that agents implementing the passwword question protocol can identify what we are asking for.

File: src/shared/cryptsetup-fido2.c [hunk 10]
```diff
@@ -129,12 +131,7 @@ int acquire_fido2_key(
...
                 const char *fido2_device,
-                const char *key_file,
-                size_t key_file_size,
-                uint64_t key_file_offset,
```
Comment (poettering) [inline src/shared/cryptsetup-fido2.c:137]: otoh dropping these three certainly makes sense afaics.

File: src/cryptenroll/cryptenroll.c [hunk 11]
```diff
@@ -161,6 +166,7 @@ static int parse_argv(int argc, char *argv[]) {
                 ARG_FIDO2_WITH_UP,
                 ARG_FIDO2_WITH_UV,
                 ARG_FIDO2_CRED_ALG,
+                ARG_UNLOCK_FIDO2_DEVICE,
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]: maybe use same order as in the options[] array below

Comment (PeterCxy) [reply src/cryptenroll/cryptenroll.c]: Looking at the git history of this enum, it seems that it has always been append-only regardless of the order in the options[] array? And I think too it would be best not to change the value of previously defined enum members?

File: src/cryptenroll/cryptenroll.c [hunk 12]
```diff
@@ -516,31 +589,18 @@ static int prepare_luks(
...
-                r = read_full_file_full(
-                                AT_FDCWD,
-                                arg_unlock_keyfile,
-                                0,
-                                SIZE_MAX,
-                                READ_FULL_FILE_SECURE|READ_FULL_FILE_WARN_WORLD_READABLE|READ_FULL_FILE_CONNECT_SOCKET,
-                                NULL,
-                                &password,
-                                &password_len);
+        if (arg_unlock_type == UNLOCK_KEYFILE) {
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]: use `switch (arg_unlock_type)` here maybe?

File: src/shared/cryptsetup-fido2.c (hunk 9 above)
Comment (PeterCxy) [reply src/shared/cryptsetup-fido2.c:133]: What name do you think we should pass in from cryptenroll? Or maybe just make it NULL in that case only?

File: src/cryptenroll/cryptenroll.c [hunk 13]
```diff
@@ -250,11 +257,35 @@ static int parse_argv(int argc, char *argv[]) {
                         break;
 
                 case ARG_UNLOCK_KEYFILE:
+                        if (arg_unlock_type > UNLOCK_PASSWORD)
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]: `>` comparisons for enums are always weird, given they indicate ordering of the enum values was important. but that#s not really the case here. You might as well just to the more obvious `!=` comparison here...

File: src/cryptenroll/cryptenroll.c [hunk 14]
```diff
@@ -250,11 +257,35 @@ static int parse_argv(int argc, char *argv[]) {
...
                         if (r < 0)
                                 return r;
+
+                        arg_unlock_type = UNLOCK_KEYFILE;
                         break;
 
+                case ARG_UNLOCK_FIDO2_DEVICE: {
+                        _cleanup_free_ char *device = NULL;
+
+                        if (arg_unlock_type > UNLOCK_PASSWORD)
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]: ditto

File: src/cryptenroll/cryptenroll.c [hunk 15]
```diff
@@ -250,11 +257,35 @@ static int parse_argv(int argc, char *argv[]) {
...
+                                                       "Multiple unlock methods specified at once, refusing.");
+
+                        if (!streq(optarg, "auto")) {
+                                device = strdup(optarg);
+                                if (!device)
+                                        return log_oom();
+                        }
+
+                        arg_unlock_type = UNLOCK_FIDO2;
+                        arg_unlock_fido2_device = TAKE_PTR(device);
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c:287]: either use `free_and_replace()` or place `assert(!arg_unlock_fido2_device)` here to indicate that you verified that the variable could not possibly be initialized before.

File: src/cryptenroll/cryptenroll.c [hunk 16]
```diff
@@ -474,6 +513,40 @@ static int check_for_homed(struct crypt_device *cd) {
         return 0;
 }
 
+static int prepare_luks_keyfile(
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]: hmm, this is a bit too generically named, for my taste. "prepare" can mean anything. And we use that verb already too often.

also, this should make clear this is about unlocking.

maybe load_unlock_keyfile()?

File: src/cryptenroll/cryptenroll.c [hunk 17]
```diff
@@ -516,31 +589,18 @@ static int prepare_luks(
...
-                                CRYPT_ANY_SLOT,
-                                vk,
-                                &vks,
-                                password,
-                                password_len);
+                goto out;
+        }
+
+        if (arg_unlock_type == UNLOCK_FIDO2) {
+                r = prepare_luks_fido2(cd, arg_unlock_fido2_device, vk, &vks);
```
Comment (poettering) [inline src/cryptenroll/cryptenroll.c]: similar hre: i think this should be called load_unlock_fido2() or so

File: src/shared/cryptsetup-fido2.c (hunk 9 above)
Comment (poettering) [reply src/shared/cryptsetup-fido2.c:133]: arg_node should be suitable.

File: src/cryptenroll/cryptenroll.c (hunk 11 above)
Comment (poettering) [reply src/cryptenroll/cryptenroll.c]: this is certainly nitpicky bikesheddy territory ;-)

let's say it's a bit chaotic ;-)

I think in general the the options[] array should probably follow the order in the --help text. And the enum order should follow the order in the options[] array. And the --help text should order things in a way that stuff that belongs together is shown together, somewhat grouped (i.e. not alphabetical).

But of course, this is not strictly enforced...

**Comment by poettering** [comment]
commit msg should probably reference #20230.

(i guess we can close #20230 once there's similar logic for unlocking via pkcs#11 tokens in place, hence just reference that issue, don't add "Closes:" for it just yet)

File: src/shared/cryptsetup-fido2.c (hunk 9 above)
Comment (PeterCxy) [reply src/shared/cryptsetup-fido2.c:133]: It seems that none of the other cryptsetup modes included the volume name / friendly name in their calls to ask_password_auto() either. I'll just leave this change out of this PR and potentially fix it in a future one.

File: src/shared/cryptsetup-fido2.c [hunk 18]
```diff
@@ -247,12 +244,14 @@ int acquire_fido2_key_auto(
                         required |= FIDO2ENROLL_UV_OMIT; /* compat with 248 */
 
                 ret = acquire_fido2_key(
-                                name,
-                                friendly_name,
+                                /* volume_name= */ NULL, /* not needed as key_file == NULL */
+                                /* friendly_name= */ NULL,
```
Comment (poettering) [inline src/shared/cryptsetup-fido2.c]: please undo this change too

Comment (poettering) [reply src/shared/cryptsetup-fido2.c]: fixed in wrong commit?

Comment (PeterCxy) [reply src/shared/cryptsetup-fido2.c]: Ah, sorry...

Comment (PeterCxy) [reply src/shared/cryptsetup-fido2.c]: Fixed.

**Comment by poettering** [comment]
lgtm

**Comment by PeterCxy** [comment]
Fixed an extra empty line in cryptenroll-fido2.c and it seems that the ubuntu CIs are failing for unrelated reasons again.

**Comment by PeterCxy** [comment]
Sorry for all the noise, last minor change: added proper error logging in `load_volume_key_fido2()` when `acquire_fido2_key_auto()` returns `-EAGAIN`, because that function expects callers to try again and/or prompt the user to try again. In cryptenroll we just need to pass this message to the user directly and abort.

**Comment by PeterCxy** [comment]
By the way, @bluca please do not set #20230 to be auto-closed when this PR is merged. That issue needs at least PKCS#11 unlocking to be also implemented in cryptenroll. This PR only deals with FIDO2, which is arguably the more common use-case but that issue does cover all types of tokens.
